flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/KinesisDataFetcher.java

```diff
@@ -584,17 +594,34 @@ private static ShardMetricsReporter registerShardMetrics(MetricGroup metricGroup
 	//  Miscellaneous utility functions
 	// ------------------------------------------------------------------------
 
+	/**
+	 * Function to map a Kinesis shard to a Flink subtask index.
+	 */
+	public interface ShardToSubtaskIndexFn {
+		/**
+		 * Function to map a Kinesis shard to a Flink subtask index.
+		 *
+		 * @param shard the shard to determine
+		 * @param totalNumberOfSubtasks total number of subtasks
+		 * @return index or hash code
+		 */
+		// TODO: extra parameter can be eliminated by creating hash function after runtime context is present
+		int getSubTaskIndex(StreamShardHandle shard, int totalNumberOfSubtasks);
```

Comment:
I actually prefer passing the `totalNumberOfSubtasks` value independently, instead of passing in `runtimeContext`. IMO, it provides more context on the nature of the assignment.

Moreover, IMO, having a factory as the API is much more complicating for the user.